### ISSUE TITLE
Tensor reshape

### ISSUE THREAD

**Issue by LouisStef** [VSainteuf/pytorch-psetae#18]
Hi, 
First of all, thanks for your impressive work!

I was wondering **why you reshape your 4D inputs (NxTxCxS) into 3D tensors (N*TxCxS)**.

**Question 1**
My guess is that you wanted to use BatchNorm1d layers (which only support 2D and 3D tensors).
Is there any deeper reason ?

**Question 2**
Would it be possible to use BatchNorm2D (which support 4D tensors) instead ? my guess is that it would not lead to the same normalization process (batch normalization performed by timestamp, instead of over the entire batch of the N*T samples), but to me, it seems OK though. What are you thoughts about it ?

**Comment by VSainteuf** [comment]
Hi @LouisStef and thanks!
No it's not a problem of BatchNorm. This reshape operation happens in the PSE because, by design, the PSE is applied in parallel to all timestamps of the input sequence. So merging the temporal and batch dimensions is an efficient way to process all time points of all samples in the batch at once.
After the PSE, the features are reshaped back to 4D to allow for temporal encoding with the TAE.  
So I would actually return the question, why would you want to keep the tensor 4D at the PSE stage ? 😁

**Comment by LouisStef** [comment]
I am trying to simplify the model architecture by removing these reshape transformations.
I don't get why the reshape operation makes the whole process faster / simpler.
Why not directly send the 4D vector (NxTxCxS) to MLP 1 (N x T x C x d1), then aggregate along the pixel dimension (N x T x C) ... ?
For me, processing the batch as a 4D tensor solves the problem more easily and is more straightforward.

**Comment by VSainteuf** [comment]
I had in mind that the MLP does not accept 4D inputs, but it seems that it does, so yes you could do that too. 
But you would still need to do some dimensions shenanigans:
 To answer your first question, yes you could use a batchnorm2d, that wouldn't mean that each date has a separate normalisation value, it would still be the same normalisation values for all pixels of all dates. Nevertheless, you would need to permute your tensor of features from NxTxCxS to NxCxTxS (so that T and S are treated as the H,W dimensions of image data).
And on the other hand, the 4D -> 3D reshape in the PSE is a simple .view()  call  without .contiguous() so it does not incur any additional computation.
So all in all, you would replace one .view by one .transpose and the computational load will be the same. 
But try it out and let me know if you find it's faster!

**Comment by LouisStef** [comment]
Got it ;)
